12272020/root/opt/joy2key/example.sh

```diff
@@ -0,0 +1,39 @@
+#!/bin/bash
+# Example whiptail dialog for ArkOS using RetroPie joy2key.py loader
+# by ridgek
+###########
+# PREFLIGHT
+###########
+RETROPIE_HELPERS="/opt/joy2key/RetroPie-Setup/scriptmodules/helpers.sh"
+
+if [ -f "${RETROPIE_HELPERS}" ]; then
+	source "${RETROPIE_HELPERS}"
+
+	# Required by joy2keyStart
+	scriptdir="/opt/joy2key/RetroPie-Setup"
+	# Arguments to pass to joy2keyStart (left, right, up, down, return, space, esc)
+	keyconfig=(kcub1 kcuf1 kcuu1 kcud1 0x0a 0x20 0x1b)
```

Comment:
How did you figure out these values for the keys?  I ask because while it may work for the rg351p, it doesn't for the rgb10 , rk2020, and oga devices.  Those will need a different set of keycodes for use with ArkOS.

Reply:
### Short answer ###
I didn't, those are the defaults from the [RetroPie helper.sh](https://github.com/RetroPie/RetroPie-Setup/blob/ac8d87759f4cad763affbba624be4fb206790ceb/scriptmodules/helpers.sh#L1150). Since joy2key.py already does auto-detection of the input device (see below), the best and most portable solution will be to set up .cfg files for each handheld. (Otherwise, we will have to write our own hacky autodetection function and pass ugly arrays like `(kcub1 kcuf1 kcuu1 kcud1 : : : : : : : : : : 0x0a : : : 0x20 : : : : : : : : : : : 0x1)` or something like that.)

### Long answer ###
The question is, what's the best way to generate those .cfgs? This is everything I've been able to figure out so far. Worst case, we will have to manually generate them, but hopefully we can fill in the missing pieces together and figure out a more elegant way!!!

**[joy2keyStart](https://github.com/RetroPie/RetroPie-Setup/blob/ac8d87759f4cad763affbba624be4fb206790ceb/scriptmodules/helpers.sh#L1143)**
This is a wrapper for calling joy2key.py which passes to it an input device ([/dev/input/js0 by default](https://github.com/RetroPie/RetroPie-Setup/blob/ac8d87759f4cad763affbba624be4fb206790ceb/scriptmodules/helpers.sh#L1154)), and an [array of curses capability names or ASCII bytecodes](https://github.com/RetroPie/RetroPie-Setup/blob/ac8d87759f4cad763affbba624be4fb206790ceb/scriptmodules/helpers.sh#L1141). Args 0-3 are the joypad's d-pad left, right, up, down. Args 4+ are the joypad's button id. 

**[joy2key.py](https://github.com/RetroPie/RetroPie-Setup/blob/master/scriptmodules/supplementary/runcommand/joy2key.py)**
The script polls the input devices and [looks for mapping files created by RetroPie](https://github.com/RetroPie/RetroPie-Setup/blob/ac8d87759f4cad763affbba624be4fb206790ceb/scriptmodules/supplementary/runcommand/joy2key.py#L82) (how RetroPie does this is the missing piece of the puzzle) in `/opt/retropie/configs/all/retroarch-joypads/` (a symlink which points to `/opt/retropie/configs/all/retroarch/autoconfig/`). These mapping files are almost identical to the ones in retroarch.cfg, but without the playerX part of the keys (eg, `input_device` instead of `input_device_p1`, `input_start_btn` instead of `input_player1_start_btn`, etc). 

*If it finds no mapping files, it [defaults to retroarch.cfg](https://github.com/ridgekuhn/arkos/blob/338d76c3912cae9cb887af10e6989e93452847a8/12272020/root/opt/joy2key/RetroPie-Setup/scriptmodules/supplementary/runcommand/joy2key.py#L90).*

So, [in my fork](https://github.com/ridgekuhn/arkos/blob/338d76c3912cae9cb887af10e6989e93452847a8/12272020/root/opt/joy2key/RetroPie-Setup/scriptmodules/supplementary/runcommand/joy2key.py#L33), we're just reading from `/home/ark/.config/retroarch/retroarch.cfg`

So, where do these .cfgs in `/opt/retropie/configs/all/retroarch-joypads/` come from???

**RetroPie Mysteries**
At some point, they're generated by RetroPie after you've configured a joypad in EmulationStation. The default es_input.cfg contains a clue. It looks like this:
```
<?xml version="1.0"?>
<inputList>
  <inputAction type="onfinish">
    <command>/opt/retropie/supplementary/emulationstation/scripts/inputconfiguration.sh</command>
  </inputAction>
</inputList>
```
So, the .cfgs in `retroarch-joypads` are generated by [inputconfiguration.sh](https://github.com/RetroPie/RetroPie-Setup/blob/master/scriptmodules/supplementary/emulationstation/inputconfiguration.sh), but looking at that script deepens the mystery. It never parses `es_input.cfg`. [Instead, it parses a file named](https://github.com/RetroPie/RetroPie-Setup/blob/ac8d87759f4cad763affbba624be4fb206790ceb/scriptmodules/supplementary/emulationstation/inputconfiguration.sh#L63) `es_temporaryinput.cfg`, and [this is the only mention of that file in the entire repository](https://github.com/RetroPie/RetroPie-Setup/search?q=es_temporaryinput)!

I thought that maybe I could just copy the RG351P's default [es_input.cfg](https://github.com/christianhaitian/arkos/files/5752526/es_input.cfg.txt) into my RetroPie's `es_temporaryinput.cfg`, run `inputconfiguration.sh`, and it would spit out some nice retroarch .cfgs for me in `retroarch-joypads/`. Unfortunately, that wasn't the case. The script runs and the output tells me it's reading those files, but then there's no new .cfgs to be found.

Crawling through the rest of the RetroPie repo, I can tell that when you begin configuring a device in EmulationStation, at some point, it calls [this function](https://github.com/RetroPie/RetroPie-Setup/blob/6f5e6c412e57d68fe3076ec49c02970cc3a7edd0/scriptmodules/supplementary/emulationstation.sh#L186) which generates that `inputAction` line before calling [this function](https://github.com/RetroPie/RetroPie-Setup/blob/2f06d303f5b9b93671ddc1df5be7a1c83979aa31/scriptmodules/supplementary/emulationstation/configscripts/emulationstation.sh#L12) which actually writes the mappings to `es_input.cfg`, but I have no idea when `es_temporaryinput.cfg` is generated, or why manually editing it and running `inputconfiguration.sh` doesn't generate some nice retroarch input .cfgs which we could take advantage of.

One final question that I didn't search too hard for an answer for, is how RetroPie lets RetroArch know to use those files in `retroarch-joypads/`.

** Solutions **
So, if we could solve the above mystery, we could integrate `inputconfiguration.sh` into ArkOS to always make sure we have good input mappings regardless of the device it's running on. 

Otherwise, we can create a folder of manually created device mappings and point `joy2key.py` at it (assuming that each handheld uses a different joypad deviceName or has similar enough mappings if they don't). Or, we could write an autodetection function with pre-defined arrays for passing to `joy2keyStart`, and get the joypad device name from parsing `evtest` output or some other method. 

Thanks for attending my TED Talk!

Reply:
One last note, `inputconfiguration.sh` calls [this script](https://github.com/RetroPie/RetroPie-Setup/blob/ac8d87759f4cad763affbba624be4fb206790ceb/scriptmodules/supplementary/emulationstation/configscripts/retroarch.sh#L380), which is actually supposed to generate the cfg file

Reply:
I've made a help request in the RetroPie forum, hopefully someone has the answer!

https://retropie.org.uk/forum/topic/29037/help-with-manually-generating-retroarch-autoconfigs-for-a-non-retropie-es-ra-project

Reply:
So I finally had a chance to try this on the rg351p.  I downloaded the most arkos12272020.zip file you created and copied the /opt contents to my test system.  I was able to use listen.sh to launch my test whiptail script but I have no controls on my end.  Was there something else that needs to be done to have controls recognized by joy2key on the rg351p?

Reply:
Hmm, I don't believe so! Here's the /opt/joy2key directory from my RG351P which is only a few minutes older than the commit I pushed before posting the video in here. Can you attach your /home/ark/.config/retroarch/retroarch.cfg so I can compare it to mine?
[joy2key.zip](https://github.com/christianhaitian/arkos/files/5753264/joy2key.zip)

Reply:
@christianhaitian Can you also attach your /etc/emulationstation/es_input.cfg? My retroarch.cfg and es_input.cfg should both be stock from the 12162020 image, but I'd like to check yours out just to make sure. 

One last thing, did you grant executable permissions to joy2key.py? (If this turns out to be the problem, I should def add an `if [ -x joy2key.py ]` to the wrapper script)

Reply:
Oh, did you install `pyudev`? My RG351P has it installed via `pip` per the convo above, I wonder if that could be the difference?

Reply:
Well how embarrassing.  I forgot to install pyudev as you pointed out.  
![image](https://user-images.githubusercontent.com/22970448/103335884-71a80380-4a44-11eb-9a58-fb919a3470ef.png)

Reply:
And interestingly enough, using your example.sh file, it works on the odroid go advance.  Although the analog joystick must be used to scroll through selections and the R button to select an option.  I'm sure that can be addressed with a little research on my end.

Reply:
I think I have an idea why. Looking at the odroidgo2_joypad device in ArkOS' es_input.cfg, I see the driver lists the dpad as "buttons" 8-11 rather than as a "hat". I don't speak Python so I haven't looked super closely at joy2key.py but I'm guessing it ignores retroarch.cfg's `input_player1_${dir}_btn` and only maps what it finds at `input_player1_${dir}` and `input_player1_${dir}_axis` to the arrow keys. I wonder if `input_player1_analog_dpad_mode` is also a factor?